rust/src/dataset/cleanup.rs

```diff
@@ -0,0 +1,697 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A task to clean up a lance dataset
+
+use std::collections::HashSet;
+
+use chrono::{DateTime, Duration, Utc};
+use futures::TryStreamExt;
+use object_store::path::Path;
+
+use crate::{
+    format::Manifest,
+    index::vector::index_dir,
+    io::{deletion_file_path, read_manifest, reader::read_manifest_indexes},
+    utils::temporal::utc_now,
+    Dataset, Error, Result,
+};
+
+#[derive(Default)]
+struct InvalidFiles {
+    unref_data_paths: HashSet<Path>,
+    unref_delete_paths: HashSet<Path>,
+    unref_index_uuids: HashSet<String>,
+    old_manifest_paths: Vec<Path>,
+}
+
+impl InvalidFiles {
+    fn new(all_files: &FilesList) -> Self {
+        Self {
+            unref_data_paths: HashSet::from_iter(all_files.data_file_paths.iter().cloned()),
+            unref_delete_paths: HashSet::from_iter(all_files.delete_file_paths.iter().cloned()),
+            unref_index_uuids: HashSet::from_iter(all_files.index_uuids.iter().cloned()),
+            old_manifest_paths: Vec::new(),
+        }
+    }
+}
+
+impl InvalidFiles {
+    fn finish(self, dataset: &Dataset) -> (Vec<Path>, Vec<Path>) {
+        let mut all_paths = self.old_manifest_paths;
+        all_paths.extend(self.unref_data_paths);
+        all_paths.extend(self.unref_delete_paths);
+        let all_dirs = Vec::from_iter(
+            self.unref_index_uuids
+                .into_iter()
+                .map(|uuid| index_dir(dataset, &uuid)),
+        );
+        (all_paths, all_dirs)
+    }
+}
+
+#[derive(Default)]
+struct FilesList {
+    data_file_paths: Vec<Path>,
+    manifest_file_paths: Vec<Path>,
+    index_uuids: HashSet<String>,
+    delete_file_paths: Vec<Path>,
+}
+
+fn join_paths(base: &Path, child: &Path) -> Path {
+    return child
+        .parts()
+        .fold(base.clone(), |joined, part| joined.child(part));
+}
+
+fn remove_prefix(path: &Path, prefix: &Path) -> Path {
+    let relative_parts = path.prefix_match(prefix);
+    if relative_parts.is_none() {
+        return path.clone();
+    }
+    Path::from_iter(relative_parts.unwrap())
+}
+
+#[derive(Debug, Clone)]
+struct CleanupTask<'a> {
+    dataset: &'a Dataset,
+    /// Cleanup all versions before this time
+    before: DateTime<Utc>,
+}
+
+const MINIMUM_CLEANUP_DAYS: i64 = 7;
+
+impl<'a> CleanupTask<'a> {
+    fn new(dataset: &'a Dataset, before: DateTime<Utc>) -> Self {
+        Self { dataset, before }
+    }
+
+    async fn run(&mut self) -> Result<()> {
+        self.validate()?;
+        let all_paths: Vec<Path> = self
+            .dataset
+            .object_store
+            .read_dir_all(&self.dataset.base)
+            .await?
+            .try_collect()
+            .await?;
+        let mut files_list: FilesList = Default::default();
+        for path in all_paths {
+            self.discover_path(&path, &mut files_list)?;
+        }
```

Comment:
To eliminate some temporary allocations, perhaps we can do:
```suggestion
        let mut files_list: FilesList = Default::default();
        let all_paths: Vec<Path> = self
            .dataset
            .object_store
            .read_dir_all(&self.dataset.base)
            .await?
            .try_for_each(|path| {
                Self::discover_path(&path, &mut files_list)
            })
            .await?;
```

Reply:
The new approach is stream based and avoids this kind of temporary allocation I think.